Add mm_pattern and bmm_pattern to serialized_patterns

aorenste opened the pull request:
Make it easier to serialize patterns by adding `pattern_matcher.gen_register_replacement()` which is like `pattern_matcher.register_replacement()` but also requires the replacement to be precompiled.

To precompile patterns (and save to disk) run:
```
torchgen/fuse_attention_patterns/gen_attention_patterns.py
```

- Updated the sfdp patterns to use `gen_register_replacement`.
- Add serialized patterns for mm_pattern and bmm_pattern (The 'misc' patterns don't serialize cleanly so can't be added).
- Updated the testing so it checked the round-trip patterns match and not just that it serialized the same way.
- Checking that the patterns round-trip properly found that the `users` field wasn't being serialized properly.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121314
* __->__ #121313



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

aorenste committed:
Add mm_pattern and bmm_pattern to serialized_patterns

[ghstack-poisoned]

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"

Add serialized patterns for mm_pattern and bmm_pattern following the method used by the sfdp patterns.

The 'misc' patterns don't serialize cleanly so can't be added.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"

Add serialized patterns for mm_pattern and bmm_pattern following the method used by the sfdp patterns.

The 'misc' patterns don't serialize cleanly so can't be added.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison reviewed (commented):
Cool! just curious, any exact compilation time increase from this ?

Would it be possible to add syntax so that it is easier to use serialization for future patterns? I'm imagining something like
```
register_replacement(
    gen_pattern_fn(fn, unique_key
```

which when run with an environment variable would serialize the pattern for you. And then subsequently would load the previously serialized pattern. Basically, the same UX as pytorch expect tests. 

cc @Chillee who i previously talked to abt this

aorenste commented:
> Cool! just curious, any exact compilation time increase from this ?

The compilation time improvement actually comes from the next PR in the stack (#121314).  I'm waiting for benchmarks to run so I can measure it.

> Would it be possible to add syntax so that it is easier to use serialization for future patterns? I'm imagining something like
> 
> ```
> register_replacement(
>     gen_pattern_fn(fn, unique_key
> ```
> 
> which when run with an environment variable would serialize the pattern for you. And then subsequently would load the previously serialized pattern. Basically, the same UX as pytorch expect tests.

Neat idea - I'll try that.

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"

Add serialized patterns for mm_pattern and bmm_pattern following the method used by the sfdp patterns.

The 'misc' patterns don't serialize cleanly so can't be added.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aorenste commented:
> >   
> > which when run with an environment variable would serialize the pattern for you. And then subsequently would load the previously serialized pattern. Basically, the same UX as pytorch expect tests.
> 
> Neat idea - I'll try that.

@eellison So I thought about this a bit more. Unless I'm missing the point of what you're saying I'm not sure it can really get much simpler than it is. The only real difference is that you run a script (gen_attention_patterns.py) instead of setting an environment variable. I'm not sure there's a lot of value in baking the contents of that script into the pytorch runtime.

Since we still need a common registry of some sort you still need to tell *something* global about which generators to probe.

eellison commented:
> The only real difference is that you run a script (gen_attention_patterns.py) instead of setting an environment variable

The difference is there's much code changes required in order to get it to work.

> Since we still need a common registry of some sort you still need to tell something global about which generators to probe

I'm not sure we need a global registry, just the generation of particular pattern needs to know if a serialized version exists. Which would be the same lookup that writes the pattern if the env variable was set.

I'm still okay landing without it but I do think it would be an improvement (and an easier sell in adoption for internal/external teams with patterns).

In either case could we add some tests along these lines: https://github.com/pytorch/pytorch/blob/main/test/inductor/test_pattern_matcher.py#L1004-L1008?

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"

Add serialized patterns for mm_pattern and bmm_pattern following the method used by the sfdp patterns.

The 'misc' patterns don't serialize cleanly so can't be added.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"

Add serialized patterns for mm_pattern and bmm_pattern following the method used by the sfdp patterns.

The 'misc' patterns don't serialize cleanly so can't be added.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aorenste commented:
@eellison Ok - updated to a form that I think is closer to what you suggested.

Also added full testing - which ended up catching a bug in the original sfdp patterns (the `users` parameter wasn't being written when it was `0`).

eellison commented:
cool - will take a look when test failures are fixed

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"

Add serialized patterns for mm_pattern and bmm_pattern following the method used by the sfdp patterns.

The 'misc' patterns don't serialize cleanly so can't be added.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"

Add serialized patterns for mm_pattern and bmm_pattern following the method used by the sfdp patterns.

The 'misc' patterns don't serialize cleanly so can't be added.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"

Add serialized patterns for mm_pattern and bmm_pattern following the method used by the sfdp patterns.

The 'misc' patterns don't serialize cleanly so can't be added.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented on torch/_inductor/constant_folding.py:
```diff
@@ -102,7 +103,8 @@ def run_node(self, node):
 
-            pytree.tree_map_only(torch.fx.Node, set_env, node.args)
+            # In-place is fine since we don't mutate
+            pytree.tree_map_only_(torch.fx.Node, set_env, node.args)
```
nit: unrelated..

aorenste replied:
True (Related only in that I discovered it while looking at the perf of the pattern matcher). I kind of figured for a single-line fix it wasn't worth a PR on its own (especially considering how long it takes to test/review a PR).

eellison commented on torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_1.py:
```diff
@@ -100,7 +100,7 @@
 view_default_4 = CallFunction(aten.view.default, expand_default_3, Ignored())
 bmm_default_1 = CallFunction(aten.bmm.default, view_default_3, view_default_4)
-_sfdp_pattern_1_inference = CallFunction(aten.view.default, bmm_default_1, Ignored())
+_sfdp_pattern_1_inference = CallFunction(aten.view.default, bmm_default_1, Ignored(), _users=0)
```
Is there a reason to add `_user=0` here ?

aorenste replied:
As mentioned in the summary - the patterns weren't being serialized properly before and the test didn't catch it. This is the output of the generator now that it's correctly serializing the pattern.

eellison commented on torch/_inductor/fx_passes/serialized_patterns/addmm_pattern.py:
```diff
@@ -0,0 +1,53 @@
+# mypy: ignore-errors
+
+# noqa: F401, E501
+# This is an auto-generated file. Please do not modify it by hand.
+# To re-generate, run:
+# cd ~/pytorch && python
+# torchgen/fuse_attention_patterns/gen_attention_patterns.py
```
Not relevant here right ?

aorenste replied:
Which part? It's auto-generated and you can use that script to regenerate it (the script is poorly named now since it regenerates ALL patterns, not just fuse_attention_patterns - maybe that's your issue?)

eellison replied:
Ah, I mistakenly thought it was only generating the attention patterns, and not yet the mm patterns. Maybe we can rename the file ?

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -229,6 +236,14 @@ def find_anchor_nodes(self, ctx: MatchContext, searched):
         if self in ctx.pattern_to_node:
             yield ctx.pattern_to_node[self]
 
+    def pattern_eq(self, other: Any) -> bool:
```
Is there a reason to define `pattern_eq` instead of `__eq__` ?

aorenste replied:
I originally defined `__eq__`.  The main reason I changed it is I didn't really want to have to implement `__hash__` since I only care about this for debugging.

eellison commented on torch/_inductor/fx_passes/misc_patterns.py:
```diff
@@ -58,7 +58,7 @@ def randperm_index_replacement(x, slice_shape):
         return torch.ops.aten._unsafe_index(x, (index,)), index
 
-    pattern = register_replacement(
+    register_replacement(
```
Doesn't look like we serialized this one

eellison reviewed (commented):
Looks great! A few comments. Also curious @Chillee's thoughts.

Might be easier to land just with addmm & others first and then update the fuse attention. In one go is fine too.

aorenste commented on test/inductor/test_pattern_matcher.py:
```diff
@@ -947,66 +943,50 @@ def fn2(inp, a, b):
+                # lets verify the serializer by ensuring the generated patterns
+                # also match (since search_fn_pattern is the serialized version
+                # of search_fn).
+                self.assertTrue(pattern.pattern_eq(search_fn_pattern))
```
Not sufficient (I found a serialization bug with `_users` by doing it this way).

The problem is if you just check the output then you're just checking that the transformation to the serialized form is consistent - not that you actually serialized the whole thing:
Imagine serializing a tuple `(A, B, C)` where the serializer dropped `B`:
- you'd serialize as `(A, C)`
- read that in as `(A, C)` 
- and then write it back out as `(A, C)`.
The serialized output is the same but it's clearly dropping something.

aorenste commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -1103,6 +1181,140 @@ def normalize_args(**kwargs):
+    def get_file_template() -> str:
+        auto_generated_msg = textwrap.dedent(
+            """\
+            # This is an auto-generated file. Please do not modify it by hand.
+            # To re-generate, run:
+            # cd ~/pytorch && python
+            # torchgen/fuse_attention_patterns/gen_attention_patterns.py
```
The message is relevant - `gen_attention_patterns` will rebuild ALL the patterns (including the non-fuse ones).  If you want I can rename that script to something else but I worried that something/someone might be referencing that script so figured it would be easier to just have that script do all the work.

aorenste commented on torch/_inductor/fx_passes/serialized_patterns/central_index.py:
```diff
@@ -1,123 +0,0 @@
-    '_sfdp_pattern_17_inference': _sfdp_pattern_17_inference,
-    '_sfdp_pattern_18_training': _sfdp_pattern_18_training,
-    '_sfdp_pattern_18_inference': _sfdp_pattern_18_inference,
-    '_sfdp_pattern_18_bs1_training': _sfdp_pattern_18_bs1_training,
```
They should be - there's just no central index any more (which is what I thought you asked for).  As long as they're listed by _get_sfdp_patterns() (which was needed for them to be generated previously) then they should still be generated.

eellison replied:
Right no central index is great. We still want a way of updating the patterns. The num_users were not changed, so I'm suspicious we are not generating them as part of this script anymore

aorenste replied:
I'm not sure why `num_users` would change (OTOH - I also don't actually know what `num_users` tracks) because the only thing that changed was how we generate the scripts not any usage of those scripts. But the patterns are definitely being updated. If I manually change it and then regen then they're being re-written:
```
(base) $ hg diff
diff --git a/torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_18.py b/torch/_inductor/fx_passes/serialized_patterns$
--- a/torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_18.py
+++ b/torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_18.py
@@ -205,6 +205,7 @@
 view_default_11 = CallFunction(aten.view.default, bmm_default_5, Ignored())
 permute_default_11 = CallFunction(aten.permute.default, view_default_11, Ignored())
 _sfdp_pattern_18_bs1_training = MultiOutputPattern([view_default_5,
+  "MANUALLY EDITED JUNK",
   permute_default_1,
   permute_default_3,
   permute_default_6,
(base) $ python torchgen/fuse_attention_patterns/gen_attention_patterns.py
(base) $ hg diff
(base) $
```

eellison reviewed (approved):
Looks great - thanks for answering my questions!

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"

Make it easier to serialize patterns by adding `pattern_matcher.gen_register_replacement()` which is like `pattern_matcher.register_replacement()` but also requires the replacement to be precompiled.

To precompile patterns (and save to disk) run:
```
torchgen/fuse_attention_patterns/gen_attention_patterns.py
```

- Updated the sfdp patterns to use `gen_register_replacement`.
- Add serialized patterns for mm_pattern and bmm_pattern (The 'misc' patterns don't serialize cleanly so can't be added).
- Updated the testing so it checked the round-trip patterns match and not just that it serialized the same way.
- Checking that the patterns round-trip properly found that the `users` field wasn't being serialized properly.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aorenste committed:
Update on "Add mm_pattern and bmm_pattern to serialized_patterns"

Make it easier to serialize patterns by adding `pattern_matcher.gen_register_replacement()` which is like `pattern_matcher.register_replacement()` but also requires the replacement to be precompiled.

To precompile patterns (and save to disk) run:
```
torchgen/fuse_attention_patterns/gen_attention_patterns.py
```

- Updated the sfdp patterns to use `gen_register_replacement`.
- Add serialized patterns for mm_pattern and bmm_pattern (The 'misc' patterns don't serialize cleanly so can't be added).
- Updated the testing so it checked the round-trip patterns match and not just that it serialized the same way.
- Checking that the patterns round-trip properly found that the `users` field wasn't being serialized properly.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

awgu commented:
I think because of this PR, I see the following when rebasing and trying to re-build PyTorch:
```
(pytorch-3.10) [andgu@devgpu011.cco1 /data/users/andgu/pytorch (unshard)]$ time TORCH_CUDA_ARCH_LIST="8.0 9.0" USE_CUDA=1 USE_DISTRIBUTED=1 python setup.py develop 2>&1 | tee build_log.txt
ninja: error: rebuilding 'build.ninja': '/data/users/andgu/pytorch/torchgen/fuse_attention_patterns/gen_attention_patterns.py', needed by 'aten/src/ATen/core_generated_declarations_yaml.cmake', missing and no known rule to make it
Building wheel torch-2.4.0a0+git288cae5
-- Building version 2.4.0a0+git288cae5
cmake --build . --target install --config Release -- -j 48
```

aorenste commented:
> I think because of this PR, I see the following when rebasing and trying to re-build PyTorch:
> 
> ```
> (pytorch-3.10) [andgu@devgpu011.cco1 /data/users/andgu/pytorch (unshard)]$ time TORCH_CUDA_ARCH_LIST="8.0 9.0" USE_CUDA=1 USE_DISTRIBUTED=1 python setup.py develop 2>&1 | tee build_log.txt
> ninja: error: rebuilding 'build.ninja': '/data/users/andgu/pytorch/torchgen/fuse_attention_patterns/gen_attention_patterns.py', needed by 'aten/src/ATen/core_generated_declarations_yaml.cmake', missing and no known rule to make it
> Building wheel torch-2.4.0a0+git288cae5
> -- Building version 2.4.0a0+git288cae5
> cmake --build . --target install --config Release -- -j 48
> ```

Can you try adding "--cmake" after `develop`? That will kick the system to regenerate cmake files.

awgu commented:
@aorenste Ah, thanks for the tip. I already ran `git clean` and am rebuilding now. Sorry for the noise!

aorenste commented:
> @aorenste Ah, thanks for the tip. I already ran `git clean` and am rebuilding now. Sorry for the noise!

From my experience I generally follow this order until one of them works:
0. Run "setup.py develop"
1. Run "setup.py develop --cmake"
2. Run "setup.py clean && setup.py develop"
3. Run "git clean && setup.py develop"
4. Throw computer out window.

xuzhao9 commented:
This PR seems to cause sam_fast CUDA memory leak: https://github.com/pytorch/benchmark/issues/2231

Is this behavior expected?

xuzhao9 commented:
> > This PR seems to cause sam_fast CUDA memory leak on torchbench sam_fast model: [pytorch/benchmark#2231](https://github.com/pytorch/benchmark/issues/2231)
> > Is this behavior change expected?
> 
> No - not expected. I've been trying to repro it but having trouble getting the benchmark to run - I'll keep trying.

Created https://github.com/pytorch/pytorch/issues/124081, please reach out there if you have any problems reproducing, thanks!
